### PR TITLE
Supporting streaming API calls to callable functions

### PR DESCRIPTION
**Pull request by taeold**
The feature is not complete, but is fully backward compatible and not a risk to existing users.

One unintended change:

Since we are introducing optional 2nd argument the v2 callable handlers, we can no longer sniff the args length to differentiate v1 and v2 handlers. This requires us to explicitly set the version when invoking the the common HTTP callable wrapper.

### CODE REVIEW

File: src/v2/providers/https.ts [hunk 1]
```diff
@@ -356,11 +357,11 @@ export function onCall<T = any, Return = any | Promise<any>>(
...
  */
 export function onCall<T = any, Return = any | Promise<any>>(
-  handler: (request: CallableRequest<T>) => Return
+  handler: (request: CallableRequest<T>, response?: CallableProxyResponse) => Return
 ): CallableFunction<T, Return extends Promise<unknown> ? Return : Promise<Return>>;
 export function onCall<T = any, Return = any | Promise<any>>(
   optsOrHandler: CallableOptions | ((request: CallableRequest<T>) => Return),
-  handler?: (request: CallableRequest<T>) => Return
+  handler?: (request: CallableRequest<T>, response?: CallableProxyResponse) => Return
 ): CallableFunction<T, Return extends Promise<unknown> ? Return : Promise<Return>> {
```
Comment (blidd-google) [inline src/v2/providers/https.ts:365]: TIL you can use the ternary operator with generics

**Review by blidd-google — APPROVED** [review]
LGTM, left some minor comments

File: src/v2/providers/https.ts (hunk 1 above)
Comment (taeold) [reply src/v2/providers/https.ts:365]: It's such weird TS feature

https://www.typescriptlang.org/docs/handbook/2/conditional-types.html

File: src/v2/providers/https.ts [hunk 2]
```diff
@@ -378,16 +379,17 @@ export function onCall<T = any, Return = any | Promise<any>>(
...
       cors: { origin, methods: "POST" },
       enforceAppCheck: opts.enforceAppCheck ?? options.getGlobalOptions().enforceAppCheck,
       consumeAppCheckToken: opts.consumeAppCheckToken,
     },
-    fixedLen
+    fixedLen,
+    "v2"
   );
 
   func = wrapTraceContext(func);
```
[the remark this answers is not in the document]
Comment (taeold) [reply src/v2/providers/https.ts:395]: oh that's an excellent catch. I think this is a bug - let's fix it in a different PR though.

File: src/common/providers/https.ts [hunk 3]
```diff
@@ -698,10 +711,15 @@ export function onCallHandler<Req = any, Res = any>(
...
+function encodeSSE(data: unknown): string {
+  return `data: ${JSON.stringify(data)}\n`;
+}
+
 /** @internal */
 function wrapOnCallHandler<Req = any, Res = any>(
   options: CallableOptions,
-  handler: v1CallableHandler | v2CallableHandler<Req, Res>
+  handler: v1CallableHandler | v2CallableHandler<Req, Res>,
+  version: "v1" | "v2"
```
[the remark this answers is not in the document]
Comment (taeold) [reply src/common/providers/https.ts:722]: ah good one. will change to `'gcfv1' | 'gcfv2'` instead.

File: src/v2/providers/https.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (taeold) [reply src/v2/providers/https.ts:395]: https://github.com/firebase/firebase-functions/pull/1634

**Pull request merged**